velox/expression/ExprOptimizer.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/ExprConstants.h"
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/expression/ExprUtils.h"
+
+namespace facebook::velox::expression {
+
+namespace {
+
+// Tries constant folding expression with tryEvaluateConstantExpression API.
+// If constant folding throws VeloxUserError, returns original expression when
+// replaceEvalErrorWithFailExpr is false, otherwise returns a fail function
+// with the error message.
+core::TypedExprPtr tryConstantFold(
+    const core::TypedExprPtr& expr,
+    core::QueryCtx* queryCtx,
+    memory::MemoryPool* pool,
+    bool replaceEvalErrorWithFailExpr) {
+  try {
+    if (auto results =
+            exec::tryEvaluateConstantExpression(expr, pool, queryCtx, false)) {
+      return std::make_shared<core::ConstantTypedExpr>(results);
+    }
+  } catch (VeloxUserError& e) {
+    if (replaceEvalErrorWithFailExpr) {
+      return std::make_shared<core::CallTypedExpr>(
+          UNKNOWN(),
+          expression::kFail,
+          std::make_shared<core::ConstantTypedExpr>(VARCHAR(), e.what()));
+    }
+  }
+  // Return the expression unevaluated.
+  return expr;
+}
+
+const core::TypedExprPtr foldInputs(
+    const core::TypedExprPtr& expr,
+    core::QueryCtx* queryCtx,
+    memory::MemoryPool* pool,
+    bool replaceEvalErrorWithFailExpr) {
+  if (expr->isCallKind()) {
+    std::vector<core::TypedExprPtr> foldedInputs;
+    for (const auto& input : expr->inputs()) {
+      const auto foldedInput =
+          optimize(input, queryCtx, pool, replaceEvalErrorWithFailExpr);
+      foldedInputs.push_back(foldedInput);
+    }
+    const auto* callExpr = expr->asUnchecked<core::CallTypedExpr>();
+
+    return std::make_shared<core::CallTypedExpr>(
+        callExpr->type(), foldedInputs, callExpr->name());
+  } else if (expr->isCastKind()) {
+    const auto foldedInput = optimize(
+        expr->inputs().at(0), queryCtx, pool, replaceEvalErrorWithFailExpr);
+    if (expr->type() == foldedInput->type()) {
```

Comment:
Are we comparing shared pointers here? Shouldn't we compare Type objects instead?